In [3]:
# In this section, you will use the Yelp reviews dataset to answer the questions below. You can
# access the dataset via this link (Section 1 data file).
# Please answer the following questions:

# 7. Based on analyzed data showcase if there are any steps that the restaurant can take
# to improve their public appeal.


In [4]:
import pandas as pd

In [5]:
df = pd.read_csv('Section 1 data.csv')

In [6]:
# some first steps to understand the data
df.head()

Review_Date                                        Review_Text  \
0  2014-07-16                                       okay...so so   
1  2014-07-16  Some people complain about the prices but Subw...   
2  2014-07-16  Had red curry chicken, the chicken itself was ...   
3  2014-07-16  Alright, I gave the restaurant manager ample t...   
4  2014-07-16  Blehhhh :/ this place shouldn't even be in bus...   

                  User_ID             Business_ID           Business_Name  \
0  6VxJJX7h36bMCCFt7URg6w  9QqLqYIwV-n1BJPjnaYv8A        Beef 'O' Brady's   
1  viGPiPuMZnV4PR_aiA3-qw  pNQwnY_q4okdlnPiR-3RBA           Empire Bagels   
2  rwehMCinfBjhZ0IbR1zFBw  shCdCHRbnY5FTMJbWl-myQ             Thai Spices   
3  VWqt5IH8fm-k9M0CKFkJzg  HpaYCM_NCauI72LLXxC6SA  Yonaka Modern Japanese   
4  Y6-0ToMhjBsm8iYEaT2meg  FC4q3hJyF8oo984xoo3RMg               808 Sushi   

    Business_Category       City State   Latitude   Longitude  \
0                Bars   Chandler    AZ  33.303847 -111.946838   
1                Food  Las Vegas    NV  36.077299 -115.297979   
2                Thai       Mesa    AZ  33.412708 -111.875803   
3  Tapas/Small Plates  Las Vegas    NV  36.114935 -115.209737   
4          Sushi Bars  Las Vegas    NV  36.052181 -115.279227   

   Avg_Business_Star_Rating  
0                       3.5  
1                       3.5  
2                       4.0  
3                       4.5  
4                       3.5

In [7]:
# 1. How many unique restaurants could be found in this data set? (Hint: Use the
# [Business_ID] column for this evaluation.)
unique_restaurants = df['Business_ID'].nunique()
print("unique_restaurants: ", unique_restaurants)

unique_restaurants:  30276


In [8]:
# 2. Which restaurant received the highest number of reviews? What about percentage-
# wise?
restaurant_reviews = df['Business_ID'].value_counts()
restaurant_reviews_max = restaurant_reviews.idxmax()
restaurant_reviews_max_count = restaurant_reviews.max()
restaurant_reviews_max_percentage = (restaurant_reviews_max_count / len(df)) * 100
print("restaurant_reviews_max: ", restaurant_reviews_max)
print("restaurant_reviews_max_count: ", restaurant_reviews_max_count)
print("restaurant_reviews_max_percentage: ", restaurant_reviews_max_percentage.round(2))

restaurant_reviews_max:  4bEjOyTaDG24SY5TxsaUNQ
restaurant_reviews_max_count:  856
restaurant_reviews_max_percentage:  0.38


In [9]:
# 3. Which cities have got at least one 5-star review in Nevada (NV) state?
cities_5_star = df[(df['State'] == 'NV') & (df['Avg_Business_Star_Rating'] == 5)]['City'].unique()
print("cities_5_star: ", cities_5_star)

cities_5_star:  ['Las Vegas' 'Henderson' 'Boulder City' 'Nellis']


In [10]:
# 4. Which city has the highest number of reviews in the Business Category of “Hotels &amp;
# Travel”? What about percentage-wise?
city_reviews = df[df['Business_Category'] == 'Hotels & Travel']['City'].value_counts()
city_reviews_max = city_reviews.idxmax()
city_reviews_max_count = city_reviews.max()
city_reviews_max_percentage = (city_reviews_max_count / len(df)) * 100
print("city_reviews_max: ", city_reviews_max)
print("city_reviews_max_count: ", city_reviews_max_count)
print("city_reviews_max_percentage: ", city_reviews_max_percentage.round(2))

city_reviews_max:  Las Vegas
city_reviews_max_count:  10245
city_reviews_max_percentage:  4.5


In [11]:
# 5. At what day of the week people are more likely to post their reviews?
df['Review_Date'] = pd.to_datetime(df['Review_Date'])
df['Review_Day'] = df['Review_Date'].dt.day_name()
day_reviews = df['Review_Day'].value_counts()
day_reviews_max = day_reviews.idxmax()
print("day_reviews_max: ", day_reviews_max)

day_reviews_max:  Monday


In [12]:
# 6. Showcase if there are any trends regarding restaurant performance as time goes by.

df['Review_Year'] = df['Review_Date'].dt.year
#lets check one by one for each year by average rating
df.groupby('Review_Year')['Avg_Business_Star_Rating'].mean()

Review_Year
2005    3.622549
2006    3.588431
2007    3.592101
2008    3.632379
2009    3.657304
2010    3.685897
2011    3.696744
2012    3.717803
2013    3.762827
2014    3.794501
Name: Avg_Business_Star_Rating, dtype: float64

That shows us average performance of the all restaurants is slightly increasing during the time interval

In [25]:
# 7. Based on analyzed data showcase if there are any steps that the restaurant can take
# to improve their public appeal.

In [42]:
%pip install textblob

Defaulting to user installation because normal site-packages is not writeable
   ---------------------------------------- 0.0/626.3 kB ? eta -:--:--
   - ------------------------------------- 30.7/626.3 kB 660.6 kB/s eta 0:00:01
   -- ------------------------------------ 41.0/626.3 kB 393.8 kB/s eta 0:00:02
   ------ ------------------------------- 102.4/626.3 kB 737.3 kB/s eta 0:00:01
   -------- ----------------------------- 133.1/626.3 kB 718.6 kB/s eta 0:00:01
   --------------- ------------------------ 235.5/626.3 kB 1.0 MB/s eta 0:00:01
   ---------------------- ----------------- 348.2/626.3 kB 1.4 MB/s eta 0:00:01
   -------------------------------- ------- 501.8/626.3 kB 1.7 MB/s eta 0:00:01
   ---------------------------------------- 626.3/626.3 kB 1.8 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


DEPRECATION: neuralplot 0.0.8 has a non-standard dependency specifier matplotlib>=3.1numpy>=1.16. pip 24.1 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of neuralplot or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063


In [44]:
from textblob import TextBlob

def analyze_sentiment(text):
    return TextBlob(text).sentiment.polarity

df['Review_Text'] = df['Review_Text'].astype(str)

# Now apply sentiment analysis
df['Sentiment'] = df['Review_Text'].apply(analyze_sentiment)

# Get basic statistics of sentiment
sentiment_summary = df['Sentiment'].describe()

# Check the distribution of ratings
rating_distribution = df['Avg_Business_Star_Rating'].value_counts().sort_index()

(count    227581.000000
 mean          0.239787
 std           0.217432
 min          -1.000000
 25%           0.115833
 50%           0.233333
 75%           0.357552
 max           1.000000
 Name: Sentiment, dtype: float64,
 Avg_Business_Star_Rating
 1.0      409
 1.5     1441
 2.0     3401
 2.5    10969
 3.0    28029
 3.5    60163
 4.0    82280
 4.5    33003
 5.0     7886
 Name: count, dtype: int64)

In [45]:
sentiment_summary

count    227581.000000
mean          0.239787
std           0.217432
min          -1.000000
25%           0.115833
50%           0.233333
75%           0.357552
max           1.000000
Name: Sentiment, dtype: float64

In [46]:
rating_distribution

Avg_Business_Star_Rating
1.0      409
1.5     1441
2.0     3401
2.5    10969
3.0    28029
3.5    60163
4.0    82280
4.5    33003
5.0     7886
Name: count, dtype: int64

In [36]:
#  Group by State and list the average rating of each state
df.groupby('State')['Avg_Business_Star_Rating'].mean().sort_values(ascending=False).head(5)

State
MN     5.000000
NY     5.000000
NC     4.000000
MLN    3.883621
EDH    3.849352
Name: Avg_Business_Star_Rating, dtype: float64

In [41]:
#  Group by State and list the average rating of each state
df.groupby('City')['Avg_Business_Star_Rating'].mean().sort_values(ascending=False).head(40)

City
Chandler-Gilbert                          5.000000
Arcadia                                   5.000000
New York                                  5.000000
Scottsdale, Phoenix, Chandler, Gilbert    5.000000
Eagan                                     5.000000
Deforest                                  5.000000
Rochester                                 5.000000
Litchfield Park                           4.500000
Stockbridge                               4.500000
Roslin                                    4.500000
Surprise Crossing                         4.500000
Fort Kinnaird                             4.500000
St. Jacobs                                4.500000
Water of Leith                            4.500000
Newbridge                                 4.500000
Victoria Park                             4.500000
Tonopah                                   4.400000
Fort Mcdowell                             4.375000
Midlothian                                4.333333
Cramond                   

In [23]:
print(" Top 25 business categories with highest average rating: ", df.groupby('Business_Category')['Avg_Business_Star_Rating'].mean().sort_values(ascending=False).head(25))

 Top 25 business categories with highest average rating:  Business_Category
Irrigation                 5.000000
Personal Chefs             5.000000
Employment Law             5.000000
Florists                   5.000000
Gay Bars                   5.000000
Bartenders                 5.000000
Graphic Design             5.000000
Hospice                    5.000000
Hypnosis/Hypnotherapy      5.000000
Marketing                  5.000000
Medical Spas               5.000000
Auto Detailing             5.000000
Party Equipment Rentals    5.000000
Permanent Makeup           5.000000
Officiants                 5.000000
Shared Office Spaces       5.000000
Propane                    5.000000
Talent Agencies            5.000000
Speech Therapists          5.000000
Religious Schools          5.000000
Public Relations           5.000000
Chimney Sweeps             5.000000
Printing Services          5.000000
Plus Size Fashion          5.000000
Photo Booth Rentals        4.941176
Name: Avg_Business_Star_

The restaurant can improve their public appeal by focusing on the business category with the highest average rating.

They could consider reallocating their business to states with top rating average, (That will increase the revenue according to game theory as well) :
1. MN     5.000000
2. NY     5.000000
3. NC     4.000000
4. MLN    3.883621
5. EDH    3.849352

In [15]:
# installing geopy
%pip install geopy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


DEPRECATION: neuralplot 0.0.8 has a non-standard dependency specifier matplotlib>=3.1numpy>=1.16. pip 24.1 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of neuralplot or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063


In [16]:
# 8. Bonus Question – Based on this data set which user had the highest cumulative
# travel distance? What distance has been covered by him/her?

# by considering the visit time and location, we can calculate the distance between two locations one by one for each user using the latitude and longitude
# and then sum up the distances for each user to get the total distance travelled by each user
# we can use the haversine formula to calculate the distance between two locations
# haversine formula: a = sin²(Δφ/2) + cos φ1 ⋅ cos φ2 ⋅ sin²(Δλ/2) c = 2 ⋅ atan2( √a, √(1−a) ) d = R ⋅ c where φ is latitude, λ is longitude, R is earth’s radius (mean radius = 6,371km)
# we can use the geopy library to calculate the distance between two locations using the haversine formula

from geopy.distance import geodesic

# lets calculate the distance between locations for each user by considering the visit time order and sum up the distances for each user 
# to get the total distance travelled by each user

# first we need to sort the data by user_id and review_date
df = df.sort_values(['User_ID', 'Review_Date'])

# then we can calculate the distance between locations for each user by considering the visit time order
df['Distance'] = 0
for i in range(1, len(df)):
    if df.iloc[i]['User_ID'] == df.iloc[i-1]['User_ID']:
        df.at[i, 'Distance'] = geodesic((df.iloc[i-1]['Latitude'], df.iloc[i-1]['Longitude']), (df.iloc[i]['Latitude'], df.iloc[i]['Longitude'])).kilometers

# then we can sum up the distances for each user to get the total distance travelled by each user
total_distance = df.groupby('User_ID')['Distance'].sum()
user_max_distance = total_distance.idxmax()
user_max_distance_value = total_distance.max()
print("user_max_distance: ", user_max_distance)
print("user_max_distance_value: ", user_max_distance_value.round(1), " km")

C:\Users\Murad\AppData\Local\Temp\ipykernel_17392\98573302.py:22: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.07670972827220254' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.at[i, 'Distance'] = geodesic((df.iloc[i-1]['Latitude'], df.iloc[i-1]['Longitude']), (df.iloc[i]['Latitude'], df.iloc[i]['Longitude'])).kilometers


user_max_distance:  iy7zsAT63v2GbaB8S_iR1A
user_max_distance_value:  15981.6  km
